[coolmasternet] General tidy up of addon

benalexau opened the pull request:
Please refer to the Git commit message for details.

The only functional improvement in this commit is the polling is much faster, as it now differentiates between an existing connection and a new connection.

cpmeister commented:
This PR is reintroducing a lot of null checker warnings that were resolved

benalexau commented:
@cpmeister These warnings are false positives. For example 4 warnings reported for `ControllerHandler` lines 210 - 235 concern `InputStream` being `null`. But it's not `null` because line 203 ensures it's connected, and if it isn't connected, line 207 guarantees to connect it or raise an `IOException`. You cannot even fix these incorrect warnings by adding an `assert` or `if (socket == null) { throw IllegalStateException(); }` before the first use.

What is the general approach used elsewhere in OpenHAB? Create a local variable that is a shadow of a field simply to perform a `null` check just to overcome an incorrect compiler warning is non-idiomatic.

cpmeister commented:
They aren't false positives in a concurrent environment. The value of a field may be changed to null by another thread after you have performed a null check on the field. The null checker doesn't know how you are using your code if you know for sure that that an NPE wouldn't occur during normal operation. The null checker is taking the safest approach by assuming that a nullable field could be null at any time. The only time when you could get away with null checking a field in traditional way would be to make that field `final`. Otherwise you would need to cache the field in a local variable and use that variable instead since the reference can't be changed by other threads.

That isn't to say that the null checker is flawless, it doesn't handle some of the standard collection utilities very well (`Map` for instance) and it can get annoying doing the workarounds for those. I'd even allow disabling the null check in certain circumstances for maps.

benalexau commented:
> They aren't false positives in a concurrent environment.

If we continue with the example I gave earlier (warnings for lines 210 - 235), I do not see any way another thread could set the `socket` field to `null` after it had been checked, as those checks and all subsequent usage occurs within a `synchronized` block (and so does every other method that accesses `socket`). 

I do concede it could be made `null` with unusual techniques such as:

* Another class reflectively acquiring the `private` field, setting its accessibility, and forcing it to `null`
* Using `Unsafe` to mutate the memory at the field offset
* Deploying with a Java agent that does bytecode modification
* Loading from a custom classloader that does bytecode modification
* Post-compilation processes that perform bytecode modification

If I've missed how `socket` could be made `null` in a standard Java deployment I would be appreciative if you could kindly show me where. I honestly believe it's a false positive, but concurrency issues are subtle and a second set of eyes is always helpful....

On that note I concede it's a big ask for the compiler to statically analyze field state for every possible control flow path, especially with the added complication of concurrent access. So if you can confirm the accepted pattern in OpenHAB is local variable copying, I'm happy to do that.

cpmeister commented:
I suspect that the null checker is too dumb to take into account synchronization so it might be very easy to produce false positives within a synchronized block. That said, there is still a standard technique that can make your example throw an NPE, all you need to do is create a subclass of ControllerHandler and override the isConnected method to always return true.

> So if you can confirm the accepted pattern in OpenHAB is local variable copying, I'm happy to do that.

Yes, that is the accepted pattern to resolve null warnings.

benalexau commented:
Out of interest I added the `final` modifier to `ControllerHandler` and it still reports the false positives. But the bigger priority is to stick to whatever standards are used in an established codebase, so I'll switch it over to variable copying. Thanks for confirming.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -162,52 +170,55 @@ public boolean isConnected() {
-                            String.format("Failed to connect to %s:%s", config.host, config.port));
+                final Socket socket = this.socket;
+                if (socket == null) {
+                    throw new IllegalStateException("Socket verified earlier");
```
This message wouldn't be very helpful for debugging

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -239,57 +250,80 @@ private void checkConnection() throws CoolMasterClientError {
         synchronized (socketLock) {
-            Socket localSocket = socket;
-            if (localSocket != null) {
+            final ControllerConfiguration cfg = getConfigAs(ControllerConfiguration.class);
```
It would be more optimal to save the ControllerConfiguration to a field during initialization so you don't need to create it each time. Creating the configuration class is a costly procedure.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/handler/HVACHandler.java:
```diff
@@ -49,107 +49,130 @@
-        } catch (CoolMasterClientError e) {
-            logger.error("Failed to set channel {} -> {}: {}", channel, command, e.getMessage());
+        } catch (final IOException ioe) {
+            logger.error("Failed to set channel {} -> {}: {}", channel, command, ioe.getMessage());
```
The error logging level should be reserved for issues catastrophic enough to threaten the operation of openHab itself. An error in a binding shouldn't qualify for this. Please read the [openHab coding guidelines](https://www.openhab.org/docs/developer/guidelines.html#f-logging) for expected use of logging in bindings.

Downgrade this to a warning instead.
```suggestion
            logger.warn("Failed to set channel {} -> {}: {}", channel, command, ioe.getMessage());
```

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/handler/HVACHandler.java:
```diff
@@ -49,107 +49,130 @@
+
+            updateStatus(ThingStatus.ONLINE);
+        } catch (final IOException ioe) {
+            logger.error("Failed to refresh", ioe);
```
```suggestion
            logger.warn("Failed to refresh", ioe);
```

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/handler/HVACHandler.java:
```diff
@@ -166,18 +189,18 @@ public void refresh() {
         FAN_NUM_TO_STR = new HashMap<>();
-        FAN_NUM_TO_STR.put("0", "l"); /* Low */
-        FAN_NUM_TO_STR.put("1", "m"); /* Medium */
-        FAN_NUM_TO_STR.put("2", "h"); /* High */
-        FAN_NUM_TO_STR.put("3", "a"); /* Auto */
-        FAN_NUM_TO_STR.put("4", "t"); /* Top */
+        FAN_NUM_TO_STR.put("0", "l"); // Low
+        FAN_NUM_TO_STR.put("1", "m"); // Medium
+        FAN_NUM_TO_STR.put("2", "h"); // High
+        FAN_NUM_TO_STR.put("3", "a"); // Auto
+        FAN_NUM_TO_STR.put("4", "t"); // Top
     }
```
Please move static fields to the top of the class.

benalexau commented:
Thanks @cpmeister for your review. I have amended the code accordingly.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/handler/HVACHandler.java:
```diff
@@ -47,137 +47,160 @@
+                    "CoolMasterNet Controller bridge not set");
+            return null;
+        }
+
+        return (ControllerHandler) bridge.getHandler();
```
Should probably also check if the ControllerHandler is null here and change thing status accordingly.

benalexau commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/handler/HVACHandler.java:
```diff
@@ -47,137 +47,160 @@
+            } else if (channel.endsWith(LOUVRE) && command instanceof StringType) {
+                controller.sendCommand(String.format("swing %s %s", uid, command));
+            } else if (command instanceof RefreshType) {
+                refresh();
```
Can you expand on what you mean here? Is there an addon that you'd recommend I take a look at for an example?

cpmeister replied:
https://github.com/openhab/openhab-addons/blob/c8479ac3ac6819d411eed7859702087b905d8285/bundles/org.openhab.binding.millheat/src/main/java/org/openhab/binding/millheat/internal/handler/MillheatRoomHandler.java#L65-L77

benalexau commented:
Thanks @cpmeister for your comments. I've implemented all of them.

As a result of the improved handling of the refresh case, would it be appropriate to reduce the default polling frequency? It's currently 5 seconds, but I noticed the Millheat addon is 120 seconds.

I also notice Millheat is using `QuantityType` with temperature units, so I added corresponding support to this addon.

cpmeister commented:
> As a result of the improved handling of the refresh case, would it be appropriate to reduce the default polling frequency? It's currently 5 seconds, but I noticed the Millheat addon is 120 seconds.

You don't need to change that unless you think there is a good reason to, different bindings will have different polling rates for various reasons like conserving power.

benalexau commented:
Thanks for the info. I'll leave it as-is. These are wired ethernet, wired input power (usually from air conditioning serial line) gateways, so should be fine with 5 second polls by default. Plus it seems worth keeping it there given how many other ways the status can change (ie sensed temperatures, air conditioner control panel, CoolRemote cloud app, Alexa etc).

benalexau commented:
@cpmeister are there any other changes you'd like made to this PR?

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/CoolMasterNetBindingConstants.java:
```diff
@@ -30,16 +29,16 @@
     public static final String SET_TEMP = "set_temp";
     public static final String FAN_SPEED = "fan_speed";
-    public static final String LOUVRE = "louvre_angle";
+    public static final String LOUVRE = "louvre";
```
Was this a bug?

benalexau replied:
Yes. The introduction of the `switch` statement `default` clause found it. I noted it in the commit log:

* Fix incorrect constant for louve channel

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -44,116 +47,163 @@
+    private void stopPoller() {
+        synchronized (scheduler) {
+            final ScheduledFuture<?> poller = this.poller;
+            if (poller != null && !poller.isCancelled()) {
```
There is no harm in cancelling an already cancelled task. Also I'm pretty sure you still want to set the poller back to null regardless.
```suggestion
            if (poller != null) {
```

benalexau replied:
Will make this change.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -44,116 +47,163 @@
+                checkConnection();
+                final String ls = sendCommand("ls");
+                if (ls.length() < LS_LINE_LENGTH) {
+                    throw new CoolMasterClientError("Invalid 'ls' response: '" + ls + "'");
```
Why not use the constructor that handles formatting for you?

benalexau replied:
I forgot I added one! :-) Fixed.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -44,116 +47,163 @@
-                out.write(command.getBytes(StandardCharsets.US_ASCII));
+                OutputStream out = socket.getOutputStream();
+                out.write(command.getBytes(US_ASCII));
                 out.write(LF);
 
```
It is good practice to make sure that you flush the output before you try to read a response.
```suggestion
                out.write(LF);
                out.flush();

```

benalexau replied:
Done.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -44,116 +47,163 @@
                 logger.trace("Sending command '{}'", command);
-                OutputStream out = localSocket.getOutputStream();
-                out.write(command.getBytes(StandardCharsets.US_ASCII));
+                OutputStream out = socket.getOutputStream();
```
If you are looking to make good general improvements to make, then you could wrap this output stream in a OutputStreamWriter so you can let the writer deal with Charset conversions for you. Should allow you to get rid of the `LF` constant you are using below.

benalexau replied:
Do you mean `BufferedWriter(new OutputStreamWriter(socket.getOutputStream(), US_ASCII))` so you get `BufferedWriter.newLine()`?

cpmeister replied:
I don't think a BufferedWriter would be very useful in this case since the buffer is so short lived. Also it seems like '\n' was used pretty intentionally rather than a system newline so I think using `BufferedWriter.newLine()` might break things.
My suggestion is to just use `new OutputStreamWriter(socket.getOutputStream(), US_ASCII)`

benalexau replied:
OK done.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -223,14 +277,14 @@ private void checkConnection() throws CoolMasterClientError {
                 // Solicit for a prompt given we haven't received one earlier
-                final OutputStream out = localSocket.getOutputStream();
+                final OutputStream out = socket.getOutputStream();
                 out.write(LF);
```
```suggestion
                out.write(LF);
                out.flush();
```

benalexau replied:
Done

cpmeister replied:
Still hasn't been done...

benalexau replied:
I've definitely done it this time!

benalexau commented:
@cpmeister All changes are now made, please feel free to take another look when you've got a moment.

cpmeister commented on bundles/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java:
```diff
@@ -44,116 +49,164 @@
-    private static final int SINK_TIMEOUT = 1000;
-    private static final int SOCKET_TIMEOUT = 2000;
+public final class ControllerHandler extends BaseBridgeHandler {
+    private static final byte LF = "\n".getBytes(US_ASCII)[0];
```
This should be doable now.
```suggestion
    private static final String LF = "\n";
```

benalexau replied:
Done

benalexau commented:
@cpmeister All done
